Blog: Repo Evaluation Criteria (#60)
incubyte/incubyte.github.io: content/blog/technical-assessment-evaluation-criteria.md

+++
authors = "Arohi Parikh"
categories = ["Careers", "Software Craftsmanship"]

> date = 2021-09-21T00:00:00Z

Reviewer: We should change this date so that it now comes in the latest blogs

Author: Done

> description = "How we evaluate our candidates"

Author: Changed to -> Our Interview Process

draft = false
image = "/images/2021/09/laptop-plant.jpg"
slug = "interview-process"
tags = ["Careers", "Software Craftsmanship", "TDD"]
title = "How we hire Software Craftspeople"
+++

Incubyte is a combined effort of its team. Working here entails bringing your knowledge, curiosity, and authenticity to the table. Our recruiting process plays a vital role here, and it is carefully designed to identify the right person whose technical prowess, culture, and mindset, all align with that of the Company.

> We care about our core values ([Why Incubyte](https://blog.incubyte.co/blog/why-incubyte/)) and are constantly working to create a diverse and nurturing workplace while enabling a culture of constant learning and feedback. A place where each team member can thrive, and be the best version of themselves while making steady progress towards personal and professional mastery. Read on to know about our hiring process.

Author: We deeply care about our core values ([Why Incubyte](https://blog.incubyte.co/blog/why-incubyte/)) and strive toward making our workplace a nurturing, diverse and inclusive environment. There exists a culture of constant learning and feedback enabling an equal opportunity for everyone to thrive professionally and be the best version of themselves they can be.

A typical interview process at Incubyte is divided into 5 stages:

1. Application
2. Pre-required learning
3. Call with the People Success Team

> 4. Technical Discussion Round

Author: Done

5. Software Craft Round

> 

Reviewer: We should either put a link to our career page here or cutshort page where people can find open positions.

> 

Author: This works?
1. Application [Apply here](https://www.incubyte.co/careers)

Here are some of the things we recommend: 
- [Test-Driven Development](https://www.youtube.com/watch?v=qkblc5WRn-U)
- [Software Craftsmanship](https://www.youtube.com/watch?v=c07uQGBZl0A)
- [Clean Code](https://www.youtube.com/watch?v=zV079g7Irks&feature=emb_imp_woyt)

This step also shines light on other skills needed to clear the interviews. For example, knowing common Git operations, hands-on knowledge of your IDE (with shortcuts), and an understanding of a unit testing framework to name a few!

## Call with the People Success Team
In this step, we aim to align on the role you are looking for and what we have to offer and see if there is synergy there. Being remote-first also means that it is important that we align not just on tech; but also on culture, attitude, and core values! Our People Success Team are here to listen to your goals and help answer any questions you may have about the Company. Ask away!

## Technical Discussion with our existing team members

> A 30-minute discussion around your technical background and journey so far. You can expect the conversation to revolve around languages, frameworks, and databases you have worked with; your prior experiences, projects, etc. This is for us to get introduced to the depth and breadth of your technical prowess. Here it is important that you have a good Wi-Fi connection since we'd want to have your video on during the discussion.

Author: Does this work? A 30-minute discussion focused on your technical background and journey so far. You can expect the conversation to revolve around languages, frameworks, and databases you have worked with; your prior experiences, projects, etc. This is for us to get introduced to the depth and breadth of your technical expertise. **_It is important that you have a good Wi-Fi connection since we'd want to have your video on during both, this technical discussion and the next hands-on pairing session._**

## Software Craft Round

> Congratulations! You have gotten to this round. Here is where all the fun happens. You will fire up your IDE with the language of your choice and a simple Hello World program pre-installed and get ready to get your hands dirty. We will solve a short Kata together as a team! Consider this round as a pairing exercise, in which an Incubyte developer will partner with you and act as a navigator. It is fine to look stuff up on the internet. We understand that having Maven or Gradle dependency coordinate on top of the code is difficult. (We google things while working on production projects ALL the time!)

Reviewer 2: ```suggestion
This round is where the fun begins! Fire up your IDE with the language of your choice, a simple Hello World program pre-installed, and get ready to get your hands dirty. We will solve a short Kata together as a team! Consider this round as a pairing exercise, in which an Incubyte Craftsperson will partner with you and act as a navigator. It is fine to look stuff up on the internet. We understand that having Maven or Gradle dependency coordinate on top of the code is difficult. (We Google things while working on production projects ALL the time - its the thought process that matters!)
```

> Congratulations! You have gotten to this round. Here is where all the fun happens. You will fire up your IDE with the language of your choice and a simple Hello World program pre-installed and get ready to get your hands dirty. We will solve a short Kata together as a team! Consider this round as a pairing exercise, in which an Incubyte developer will partner with you and act as a navigator. It is fine to look stuff up on the internet. We understand that having Maven or Gradle dependency coordinate on top of the code is difficult. (We google things while working on production projects ALL the time!)

Reviewer: "Hello, world program pre-installed" does not sound right. 
Should it be "A project and IDE setup, able to execute a simple hello world program"

This brings us to the end of our interview process, expect to walk away with an offer once you have cleared all these rounds. 

> Now comes the important part, how to clear this interview round? We have broken down our requirements into 2 parts:

Reviewer 2: Is this meant to be a part of the Software Craft section or is it a new one? If new, then maybe you want to call it out separately. 

Also, a general rule while writing - any single-digit number should be spelled out and double-digit should be the number (eg. two vs 12)

Author: This brings us to the end of our interview process. Expect to walk away with an offer once you have cleared all these rounds. We've even put together some additional information to help you crack it! You're welcome. ;)

**_# Cheat Sheet (a.k.a how to crack interviews at Incubyte)

We have broken this down into 2 parts:
1. Technical
2. Non-technical_**

1. Technical
2. Non-technical  

## Technical

> At this point we're trying to get to know your technical self better. How you code, think and generally vibe with the things that matter to us will be taken into consideration here. Let's start:

Author: We want to see how you code, think and generally vibe with the things that matter to us. Let's start:

  
### Build tools like Maven or Gradle

We believe that build tools are first-class members of our development life cycle. Choosing the best build and dependency management tool when beginning a new project is vital and can go a long way.

### Clean coding practices

> It is not possible to evaluate everything during an interview, but below mentioned are a few key indicators that give us an insight into how you work:

Reviewer 2: ```suggestion
It is not possible to evaluate everything during an interview, but here are a few key indicators that give us an insight into how you work.
```

#### Code format

Code writing skills are essential, but so is code readability. A neatly formatted code is readable.

#### Amount of nesting in your code

Excessive nesting makes the code difficult to understand and raises cognitive complexity.

> #### Applying best practices to your code:

Reviewer 2: ```suggestion
#### Applying best practices to your code
```

> Using clean code practices speaks much more of a personal habit as much as your coding skill. It can range from using the appropriate naming conventions to Page Object Models if you are working on test automation. Having some linters integrated into your IDE can help you adhere to best practices.

Reviewer 2: This seems like a very short list of items that we call 'best practices'. I don't think there's any way to make it a comprehensive list, but in that case, do we want to make this section more open-ended?

Author: Using clean code practices speaks much more of a personal habit as much as your coding skill. It includes - using the appropriate naming conventions, Page Object Models if you are working on test automation, and not having commented out code, to name a few. Integrating some linters into your IDE can aid you in adhering to best practices.

> Using clean code practices speaks much more of a personal habit as much as your coding skill. It can range from using the appropriate naming conventions to Page Object Models if you are working on test automation. Having some linters integrated into your IDE can help you adhere to best practices.

Reviewer: Should we discuss evaluation parameters when explaining the process?

Author: How about we call this whole section: 

**_ # Cheat Sheet (a.k.a how to crack interviews at Incubyte)

We have broken this down into two parts:

1. Technical
2. Non-technical_**

> #### Commented-out code

Reviewer: Too detailed and part of clean code already. I would consider removing all these details from this section. Also, headings are not consistent in sizes

<img width="1104" alt="image" src="https://user-images.githubusercontent.com/1727880/174306207-db773e01-3eca-4181-b371-59695d9e1d63.png">

Commented code is as good as a dead code. Version control systems, as well as modern IDEs, are great at maintaining history, code should not be kept commented with your current production code.

> ### Other attitudes we recommend

Reviewer 2: I don't know if this header is required. Also 'value' is probably a better word than 'recommend'

#### XP Practices

> TDD, Pairing, Refactoring, and working in iterations are a few practices we would love people to try. We take TDD very seriously.

Reviewer 2: ```suggestion
TDD, pairing, refactoring, and working in iterations are a few practices we value deeply and use on a daily basis. We take TDD very seriously.
```

> We do not expect everyone to be proficient at it, but learning and knowing about it will be a great plus. Preparing for the interview by doing a few Katas will help a lot. BDD is a preferred method if you are programming for Automation Testing.

Reviewer 2: ```suggestion
We do not expect everyone to be proficient at all these practices, but knowing about the benefits and committing to master them is a great start. Preparing for the interview by implementing these practices on a few Katas will help a lot. BDD is a preferred method if you are programming for Automation Testing.
```

> ## The key pointers when we hunt for the best minds are

Reviewer 2: This entire section seems redundant and also not comprehensive. Should we remove it?

- Code should be logically correct
- It’s not just about completing the task, it’s more about the approach
- We are huge fans of clean code practices
- Simplicity and readability top the game
- Effective communication and the right attitude are as vital as technical proficiency

### Non-technical

> Attitude, communication, the ability to take and give feedback, how well you can use your IDE (read-shortcuts), the ability to work collaboratively as a team, and the consultant mindset to name a few, are some qualities we look for in a future team member. A common thread that binds every single person at Incubyte is the passion to keep learning. We learn -> apply -> again learn -> again apply towards the goal of achieving professional mastery. We believe that everyone is the owner of their career and a lifelong learner. We have curated a blog on books you must read to this effect ([here](https://blog.incubyte.co/blog/books-we-believe-should-be-on-your-bookshelf-and-read-them-too/)) 

Reviewer 2: ```suggestion
Attitude, effective communication, the ability to accept and deliver feedback, how well you use your IDE (read-shortcuts), the ability to work collaboratively as a team, and a consultant mindset to name a few, are some qualities we look for in a future team member. A common thread that binds every single person at Incubyte is the passion to keep learning. We learn -> apply -> learn again -> again apply towards the goal of achieving professional mastery -> repeat perpetually. We believe that each individual owns their own career and is a lifelong learner. We have curated a blog on books you must read to this effect ([here](https://blog.incubyte.co/blog/books-we-believe-should-be-on-your-bookshelf-and-read-them-too/)) 
```

Life cannot be all work. Watch movies or binge on Netflix. Read your favorite books, write blogs if you prefer that, or even help with documents on GitHub if you want to. They are all an essential part of how we spend our time!

> ## To summarize it

Reviewer 2: ```suggestion
## To sum it all up
```

The entire interview process is carefully curated to help us identify the interviewees’ technical skills as well as more intrinsic qualities such as principles, beliefs, and values.

> We look for a thirst for knowledge and a passion for mastering your craft. If you think you have the perfect blend of skills and expertise for this, feel free to reach us at careers@incubyte.co

Reviewer 2: ```suggestion
We look for a thirst for knowledge and a passion for mastering your craft. If you think you have the perfect blend of skills and expertise for this, drop us a note on careers@incubyte.co
```

> > Find our current vacancies [here](https://incubyte.co/careers).

Reviewer 2: ```suggestion
> Our [current vacancies](https://incubyte.co/careers).
```
